<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>sicomore — sicomore • SIComORe</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="sicomore — sicomore" />
<meta property="og:description" content="Selection of Interaction effects in COmpressed  Multiple Omics REpresentation
Selection of Interaction effects in COmpressed  Multiple Omics REpresentation
From a set of input matrices and phenotype related to the same set of individual, sicomore is a two-step method which
1. find and select groups of correlated variables in each input matrix which are good predictors for the common phenotype
2. find the most predictive interaction effects between the set of data by testing for interaction between the selected groups of each input matrix" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SIComORe</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a>
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/singleData.html">Analysis of single data set</a>
    </li>
    <li>
      <a href="../articles/twoDataInteraction.html">Analysis of Interacting Data sets</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/fguinot/sicomore-pkg">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>sicomore</h1>
    
    <div class="hidden name"><code>sicomore.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Selection of Interaction effects in COmpressed  Multiple Omics REpresentation</p>
<p>Selection of Interaction effects in COmpressed  Multiple Omics REpresentation</p>
<p>From a set of input matrices and phenotype related to the same set of individual, sicomore is a two-step method which
1. find and select groups of correlated variables in each input matrix which are good predictors for the common phenotype
2. find the most predictive interaction effects between the set of data by testing for interaction between the selected groups of each input matrix</p>
    </div>

    <pre class="usage"><span class='fu'>sicomore</span>(
  <span class='no'>y</span>,
  <span class='no'>X.list</span>,
  <span class='kw'>compressions</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span>(<span class='st'>"mean"</span>, <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>X.list</span>)),
  <span class='kw'>selection</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"rho-sicomore"</span>, <span class='st'>"sicomore"</span>, <span class='st'>"mlgl"</span>),
  <span class='kw'>cuts</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span>(<span class='no'>X.list</span>, <span class='kw'>function</span>(<span class='no'>X</span>) <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/Round.html'>round</a></span>(<span class='fl'>10</span>^<span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log10</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span>(<span class='no'>X</span>)), <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log10</a></span>(<span class='fl'>2</span>), <span class='kw'>len</span> <span class='kw'>=</span>
    <span class='fl'>100</span>)))),
  <span class='kw'>choice</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"lambda.min"</span>, <span class='st'>"lambda.min"</span>),
  <span class='kw'>method.clus</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"ward.D2"</span>, <span class='st'>"ward.D2"</span>),
  <span class='kw'>depth.cut</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>3</span>, <span class='fl'>3</span>),
  <span class='kw'>mc.cores</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>taxonomy</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>stab</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>stab.param</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='kw'>B</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>100</span>, <span class='fl'>100</span>), <span class='kw'>cutoff</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>0.75</span>, <span class='fl'>0.75</span>), <span class='kw'>PFER</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>1</span>, <span class='fl'>1</span>))
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>y</th>
      <td><p>response variable (phenotype)</p></td>
    </tr>
    <tr>
      <th>X.list</th>
      <td><p>a list of input matrices. Must have the same number of rows (number of observations). May have different number of columns (predictors)</p></td>
    </tr>
    <tr>
      <th>compressions</th>
      <td><p>a vector of string for the compression methods for each data sets. Default used the mean to compressed predictors in the same group.</p></td>
    </tr>
    <tr>
      <th>selection</th>
      <td><p>a string for the method used for variable selection for each data set.
Specify "sicomore" to use the method specifically developed for the package, "hcar" to use the method developped by Park et al. (2007)
 or
"mlgl" for the method of Grimonprez (2016)
.</p></td>
    </tr>
    <tr>
      <th>cuts</th>
      <td><p>a list of numeric vector defining the cut levels to be considered for each hierarchy. By default a sequence of 100 levels is used.</p></td>
    </tr>
    <tr>
      <th>choice</th>
      <td><p>a string (either "lambda.min" or "lambda.1se"). Indicates how the tuning parameter is chosen in the penalized regression approach.</p></td>
    </tr>
    <tr>
      <th>method.clus</th>
      <td><p>a vector of string specifying the method used for the hierarchical clustering, one for each input matrix in X.list.
By default, the hierarchy is obtain by a WARD clustering on the scaled input matrix.
To use an SNP-specific spatially contrained hierarchical clustering (Dehman et al. 2015)
 from package adjclust, specify "snpClust".
It is also possible to specify no hierarchy with "noclust".</p></td>
    </tr>
    <tr>
      <th>depth.cut</th>
      <td><p>a vector of integers specifying the depth of the search space for the variable selection part of the algorithm.
This argument allows to increase the speed of the algorithm by restraining the search space without affecting too much the performance.
A value between 3 and 6 is recommended, the smaller the faster.</p></td>
    </tr>
    <tr>
      <th>mc.cores</th>
      <td><p>an integer for the number of cores to use in the parallelization of the cross-validation and some other functions.</p></td>
    </tr>
    <tr>
      <th>taxonomy</th>
      <td><p>a hierarchical tree object constructed using taxonomical unit.
This argument could be useful if one want to provid a specific hierarchical tree based on taxonomy rather than euclidean distance.
Recommended for those who want to detect genomic-metagenomic interactions.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>not yet documented</p></td>
    </tr>
    <tr>
      <th>stab</th>
      <td><p>not yet documented</p></td>
    </tr>
    <tr>
      <th>stab.param</th>
      <td><p>not yet documented</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a RC object with class <code>sicomore</code> with methods <code><a href='https://rdrr.io/r/graphics/plot.html'>plot()</a></code>, <code>getSignificance()</code> and the following fields</p><ul>
<li><p>pval:A matrix of p-values for each interactions effects between the compressed variables originating from 2 input matrices.</p></li>
<li><p>pval.beta1:A matrix of p-values for the corresponding main effects of the compressed variables originating from the first input matrix</p></li>
<li><p>pval.beta2:A matrix of p-values for the corresponding main effects of the compressed variables originating from thesecond input matrix</p></li>
<li><p>modelsA class'sicomore-model' RC object obtain from <code>getHierLevel</code> function and with methods <code>nGrp()</code>, <code>nVar()</code>, <code>getGrp()</code>, <code>getVar()</code>,
 <code>getCV()</code>, <code>getX.comp()</code>, <code>getCoef()</code></p></li>
<li><p>tuplets:A list of integer vector specifying the indexes of the compressed variables which are fitted together in a linear model with interaction.</p></li>
<li><p>dim: A array of integers specifying the dimension of the compressed matrices.</p></li>
</ul>

    <h2 class="hasAnchor" id="functions-in-sicomore"><a class="anchor" href="#functions-in-sicomore"></a>Functions in sicomore</h2>

    

<p>TODO</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Dehman A, Ambroise C, Neuvial P (2015).
&#8220;Performance of a blockwise approach in variable
                  selection using linkage disequilibrium information.&#8221;
<em>BMC Bioinformatics</em>, <b>16</b>, 148.<br /><br /> Grimonprez Q (2016).
<em>Selection de groupes de variables corrÃ©lÃ©es en grande dimension</em>.
Ph.D. thesis, Université de Lille.<br /><br /> Park MY, Hastie T, Tibshirani R (2007).
&#8220;Averaged gene expressions for regression.&#8221;
<em>Biostatistics</em>, <b>8</b>(2), 212--227.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#functions-in-sicomore">Functions in sicomore</a></li>
      <li><a href="#references">References</a></li>
    </ul>

    <h2>Author</h2>
    <p>Julien Chiquet <a href='mailto:julien.chiquet@inrae.fr'>julien.chiquet@inrae.fr</a></p>
<p>Florent Guinot <a href='mailto:fly.guinot@gmail.com'>fly.guinot@gmail.com</a></p>
<p>Marie Szafranski  <a href='mailto:marie.szafranski@math.cnrs.fr'>marie.szafranski@math.cnrs.fr</a></p>
<p>Christophe Ambroise <a href='mailto:christophe.ambroise@genopole.cnrs.fr'>christophe.ambroise@genopole.cnrs.fr</a></p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Christophe Ambroise, Julien Chiquet, Florent Guinot, Marie Szafranski.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


