<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of two related data sets with SIComORe • sicomore</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of two related data sets with SIComORe">
<meta property="og:description" content="sicomore">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sicomore</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1-9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a>
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/singleData.html">Analysis of a single data set</a>
    </li>
    <li>
      <a href="../articles/twoDataInteraction.html">Analysis of two onteracting data sets</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/fguinot/sicomore-pkg">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysis of two related data sets with SIComORe</h1>
            <h3 class="subtitle">Detect relevant interactions between structured data</h3>
                        <h4 class="author">SICoMORe team</h4>
            
            <h4 class="date">2020-03-26</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jchiquet/sicomore-pkg/blob/master/vignettes/twoDataInteraction.Rmd"><code>vignettes/twoDataInteraction.Rmd</code></a></small>
      <div class="hidden name"><code>twoDataInteraction.Rmd</code></div>

    </div>

    
    
<p>This vignette illustrates how SIComORe can be used to analyze two interacting data sets whose respective sets of predictors are related jointly to a response variable. This may occur for instance when metagenomic data and genetic data are jointly used to predict a phenotype of interest: a group of OTU in the metagenomic data and a group of SNP in the genetic data may predict well jointly a response.</p>
<div id="requirements" class="section level2">
<h2 class="hasAnchor">
<a href="#requirements" class="anchor"></a>Requirements</h2>
<p>On top of <code>sicomore</code>, this will require a couple of standard packages tailored for data manipulation, representation or benchmarking:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1234</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">sicomore</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyverse</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">corrplot</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">microbenchmark</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">mvtnorm</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Matrix</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">reshape2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">cowplot</span>)</pre></body></html></div>
</div>
<div id="statistical-model" class="section level2">
<h2 class="hasAnchor">
<a href="#statistical-model" class="anchor"></a>Statistical Model</h2>
<p>SIComORe assumes an underlying linear model to link a phenotype (or response) of interest to a set several matrices of correlated predictors (see <span class="citation">Ambroise et al. (<a href="#ref-sicomore">2018</a>)</span>). In the case at hand, we consider two matrices (says, <span class="math inline">\(\mathbf{X}^{(1)}\)</span> for the genetic data and <span class="math inline">\(\mathbf{X}^{(2)}\)</span> for the metagenomic data). The generative model is thus</p>
<p><span class="math display">\[\begin{equation*}
  \mathbf{y} = \sum_{g \in \mathcal{G}^{(1)}} \phi( \mathbf{X}^{(1)}_{g}) \beta^{(1)}_g + \sum_{h \in \mathcal{G}^{(2)}} \phi( \mathbf{X}^{(2)}_{h}) \beta^{(2)}_h + \sum_{g \in \mathcal{G}^{(1)}} \sum_{h \in \mathcal{G}^{(2)}} \phi( \mathbf{X}^{(1)}_{G}) \phi( \mathbf{X}^{(2)}_{h}) \beta_{g h} + \boldsymbol\varepsilon, \qquad \boldsymbol\varepsilon \sim \mathcal{N}(\mathbf{0}_n, \mathbf{I}_n \sigma^2),
\end{equation*}\]</span> where</p>
<ul>
<li>
<span class="math inline">\(\mathbf{y}\)</span> is a vector of phenotype <span class="math inline">\(n\)</span> observation,</li>
<li>
<span class="math inline">\(\boldsymbol\varepsilon\)</span> is a size-<span class="math inline">\(n\)</span> Gaussian vector of noise,</li>
<li>
<span class="math inline">\(\mathbf{X}^{(1)}, \mathbf{X}^{(2)}\)</span> are genetic and metagenomic matrices of predictors with <span class="math inline">\(n\)</span> rows and <span class="math inline">\(p^{(1)}, p^{(2)}\)</span> columns,</li>
<li>
<span class="math inline">\(\mathcal{G}^{(1)}, \mathcal{G}^{(2)}\)</span> are latent group structures with <span class="math inline">\(K^{(1)}, K^{(2)}\)</span> groups,</li>
<li>
<span class="math inline">\(\phi\)</span> is a function (typically the mean “in row”) for compressing <span class="math inline">\(\mathbf{X}^{(1)}_g, \mathbf{X}^{(2)}_g\)</span> into a vector,</li>
<li>
<span class="math inline">\(\boldsymbol\beta^{(1)}, \boldsymbol\beta^{(2)}\)</span> are vectors of main effects with respectively <span class="math inline">\(K^{(1)},K^{(2)}\)</span> coefficients,</li>
<li>
<span class="math inline">\(\boldsymbol\beta\)</span> is a vector of interaction effects with respectively <span class="math inline">\(K^{(1)} \times K^{(2)}\)</span> coefficients.</li>
</ul>
</div>
<div id="data-sets-generations" class="section level2">
<h2 class="hasAnchor">
<a href="#data-sets-generations" class="anchor"></a>Data sets generations</h2>
<div id="group-structured-matrices-of-predictors" class="section level3">
<h3 class="hasAnchor">
<a href="#group-structured-matrices-of-predictors" class="anchor"></a>Group-structured Matrices of predictors</h3>
<p>Data generation goes pretty much just like in the example on an single data set (see <a href="singleData.html">the corresponding vignette</a> for details), but here two data sets are required.</p>
<p>These two data sets share a common sample size:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">n</span> <span class="kw">&lt;-</span> <span class="fl">100</span></pre></body></html></div>
<p>The first data set is composed by 50 variables spread into 5 groups approximately balanced:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">p1</span> <span class="kw">&lt;-</span> <span class="fl">50</span> <span class="co"># number of variables </span>
<span class="no">K1</span> <span class="kw">&lt;-</span> <span class="fl">5</span>  <span class="co"># number of groups</span>
<span class="no">grp1_size</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Multinom.html">rmultinom</a></span>(<span class="fl">1</span>,<span class="no">p1</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="no">p1</span>/<span class="no">K1</span>,<span class="no">K1</span>)) <span class="co">## group sizes</span>
<span class="no">Sigma1</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="no">K1</span>,<span class="fl">.5</span>,<span class="fl">.95</span>) <span class="kw">%&gt;%</span> <span class="co"># a vector defining within-group correlation</span>
  <span class="fu">map2</span>(<span class="no">grp1_size</span>, ~ <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="no">.x</span>, <span class="no">.y</span>, <span class="no">.y</span>)) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/bdiag.html">bdiag</a></span>()
<span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span>(<span class="no">Sigma1</span>) <span class="kw">&lt;-</span> <span class="fl">1</span>
<span class="no">X1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mvtnorm/man/Mvnorm.html">rmvnorm</a></span>(<span class="no">n</span>, <span class="kw">sigma</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">Sigma1</span>))</pre></body></html></div>
<p>The second data set is composed by 40 variables spread into 10 groups approximately balanced:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">p2</span> <span class="kw">&lt;-</span> <span class="fl">40</span>
<span class="no">K2</span> <span class="kw">&lt;-</span> <span class="fl">10</span>
<span class="no">grp2_size</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Multinom.html">rmultinom</a></span>(<span class="fl">1</span>,<span class="no">p2</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="no">p2</span>/<span class="no">K2</span>,<span class="no">K2</span>)) <span class="co">## group sizes</span>
<span class="no">Sigma2</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="no">K2</span>,<span class="fl">.5</span>,<span class="fl">.95</span>) <span class="kw">%&gt;%</span> <span class="co"># a vector defining within-group correlation</span>
  <span class="fu">map2</span>(<span class="no">grp2_size</span>, ~ <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="no">.x</span>, <span class="no">.y</span>, <span class="no">.y</span>)) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/bdiag.html">bdiag</a></span>()
<span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span>(<span class="no">Sigma2</span>) <span class="kw">&lt;-</span> <span class="fl">1</span>
<span class="no">X2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mvtnorm/man/Mvnorm.html">rmvnorm</a></span>(<span class="no">n</span>, <span class="kw">sigma</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">Sigma2</span>))</pre></body></html></div>
<p>The matrices of empirical correlations of <span class="math inline">\(\mathbf{X}^{(1)}\)</span> and <span class="math inline">\(\mathbf{X}^{(2)}\)</span> exhibit the block structures of each data set:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>))
<span class="fu"><a href="https://rdrr.io/pkg/corrplot/man/corrplot.html">corrplot</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="no">X1</span>), <span class="kw">method</span> <span class="kw">=</span> <span class="st">"color"</span>, <span class="kw">tl.pos</span> <span class="kw">=</span> <span class="st">"n"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/corrplot/man/corrplot.html">corrplot</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="no">X2</span>), <span class="kw">method</span> <span class="kw">=</span> <span class="st">"color"</span>, <span class="kw">tl.pos</span> <span class="kw">=</span> <span class="st">"n"</span>)</pre></body></html></div>
<p><img src="twoDataInteraction_files/figure-html/covariance%20matrices-1.png" width="768"></p>
</div>
<div id="phenotyperesponse-vector" class="section level3">
<h3 class="hasAnchor">
<a href="#phenotyperesponse-vector" class="anchor"></a>Phenotype/Response vector</h3>
<p>The phenotype is a linear combinaison of the <em>compressed</em> version of the predictors. Here, we use the mean to compressed a group of variables:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">grp1_ind</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">1</span>:<span class="no">K1</span>, <span class="no">grp1_size</span>) <span class="co"># vector of group indices</span>
<span class="no">grp2_ind</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">1</span>:<span class="no">K2</span>, <span class="no">grp2_size</span>) <span class="co"># vector of group indices</span>
<span class="no">main_effects_1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/rowsum.html">rowsum</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="no">X1</span>), <span class="no">grp1_ind</span>)/<span class="fu"><a href="https://rdrr.io/r/base/tabulate.html">tabulate</a></span>(<span class="no">grp1_ind</span>))
<span class="no">main_effects_2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/rowsum.html">rowsum</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="no">X2</span>), <span class="no">grp2_ind</span>)/<span class="fu"><a href="https://rdrr.io/r/base/tabulate.html">tabulate</a></span>(<span class="no">grp2_ind</span>))</pre></body></html></div>
<p>This time we also need the matrix of data interactions, with size <span class="math inline">\(n \times (K^{(1)} \times K^{(2)})\)</span>.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">interactions</span> <span class="kw">&lt;-</span>
  <span class="fu">map</span>(<span class="fl">1</span>:<span class="no">K2</span>, ~ <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span>(<span class="no">main_effects_1</span>, <span class="fl">1</span>, <span class="no">main_effects_2</span>[, <span class="no">.x</span>], <span class="st">"*"</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="no">cbind</span>, <span class="no">.</span>)</pre></body></html></div>
<p>The parameters <span class="math inline">\(\boldsymbol\beta^{(1)}, \boldsymbol\beta^{(2)}\)</span> and <span class="math inline">\(\boldsymbol\beta\)</span> in the linear model originating the response variable are chosen such that there is <span class="math inline">\(s=2\)</span> non-null interactions in the matrix <span class="math inline">\(\beta\)</span>:</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">s</span> <span class="kw">&lt;-</span> <span class="fl">2</span> <span class="co"># number of non-null interactions</span>
<span class="no">theta</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html">Matrix</a></span>(<span class="fl">0</span>, <span class="no">K1</span>, <span class="no">K2</span>)
<span class="no">theta</span>[<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fl">1</span>:(<span class="no">K1</span> * <span class="no">K2</span>), <span class="no">s</span>)] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="no">s</span>, <span class="fl">2</span>, <span class="fl">4</span>)</pre></body></html></div>
<p>The main effects associated with these interactions are drawn different from zero:</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">beta1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">0</span>, <span class="no">K1</span>)
<span class="no">beta2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">0</span>, <span class="no">K2</span>)
<span class="no">ind</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">theta</span> <span class="kw">!=</span><span class="fl">0</span> , <span class="kw">arr.ind</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="no">beta1</span>[<span class="no">ind</span>[, <span class="fl">1</span>]] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="no">s</span>, <span class="fl">1</span>, <span class="fl">2</span>)
<span class="no">beta2</span>[<span class="no">ind</span>[, <span class="fl">2</span>]] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="no">s</span>, <span class="fl">1</span>, <span class="fl">2</span>)</pre></body></html></div>
<p>Finally we can draw the response vector, with a noise level aiming for coefficient of determination of 0.75 on average.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">sigma</span> <span class="kw">&lt;-</span> <span class="fl">3</span>
<span class="no">epsilon</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="no">n</span>) * <span class="no">sigma</span>
<span class="no">mu</span> <span class="kw">&lt;-</span> <span class="fl">3</span>
<span class="no">y</span> <span class="kw">&lt;-</span> <span class="no">mu</span> + <span class="no">interactions</span> <span class="kw">%*%</span> <span class="fu">as</span>(<span class="no">theta</span>, <span class="st">'sparseVector'</span>) + <span class="no">main_effects_1</span> <span class="kw">%*%</span> <span class="no">beta1</span> + <span class="no">main_effects_2</span> <span class="kw">%*%</span> <span class="no">beta2</span> + <span class="no">epsilon</span>
<span class="no">y</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="no">y</span>)</pre></body></html></div>
</div>
</div>
<div id="adjusting-multiple-hierarchies" class="section level2">
<h2 class="hasAnchor">
<a href="#adjusting-multiple-hierarchies" class="anchor"></a>Adjusting multiple hierarchies</h2>
<p>We try to recover the correct levels of compression only from the two original input matrices of predictors and the response vector. The main function <code>sicomore</code> uses by default a Ward criteria for both hierarchical clustering. At each pairs of levels in the hierarchies, a LASSO criterion is used to perform variables selection. Cross-validation is used to tune the sparisty level. All these defaults can be controlled byy the user:</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">res</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/sicomore.html">sicomore</a></span>(
            <span class="kw">y</span>            <span class="kw">=</span> <span class="no">y</span>,            <span class="co"># response vector</span>
            <span class="kw">X.list</span>       <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">X1</span>, <span class="no">X2</span>), <span class="co"># list of matrices of predictors</span>
            <span class="kw">choice</span>      <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"lambda.min"</span>, <span class="st">"lambda.min"</span>),
            <span class="kw">method.clus</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ward.D2"</span>,<span class="st">"ward.D2"</span>)
        )</pre></body></html></div>
<pre><code>## 
## Considering hierarchy 1  -  rho-sicomore selection method.
## Considering hierarchy 2  -  rho-sicomore selection method.</code></pre>
<p>We can plot the most striking interactions with the method <code><a href="https://rdrr.io/r/graphics/plot.html">plot()</a></code> and compare with the “true” model:</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">theta_full</span> <span class="kw">&lt;-</span> <span class="no">theta</span>[<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">1</span>:<span class="no">K1</span>, <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">grp1_ind</span>)), <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">1</span>:<span class="no">K2</span>, <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">grp2_ind</span>))] <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()
<span class="no">plot_true</span> <span class="kw">&lt;-</span>
  <span class="kw pkg">reshape2</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span>(<span class="no">theta_full</span>) <span class="kw">%&gt;%</span>
  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="no">Var1</span>, <span class="no">Var2</span>, <span class="kw">fill</span><span class="kw">=</span><span class="no">value</span>)) +
  <span class="fu">geom_tile</span>(<span class="kw">show.legend</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">colour</span> <span class="kw">=</span> <span class="st">"grey80"</span>) +
  <span class="fu">scale_fill_gradient2</span>(<span class="kw">high</span> <span class="kw">=</span> <span class="st">"grey40"</span>) +
  <span class="fu">theme_minimal</span>() + <span class="fu">ggtitle</span>(<span class="st">"Matrix of true interactions"</span>) +
  <span class="fu">labs</span>(<span class="kw">x</span> <span class="kw">=</span> <span class="st">'Variables in X^1'</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="st">'Variables in X^2'</span>) + <span class="fu">coord_fixed</span>()
<span class="no">plot_hat</span> <span class="kw">&lt;-</span> <span class="no">res</span>$<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">main</span> <span class="kw">=</span> <span class="st">"Estimated matrix of interactions"</span>, <span class="kw">threshold</span><span class="kw">=</span><span class="fl">.05</span>)
<span class="kw pkg">cowplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(<span class="no">plot_hat</span>,<span class="no">plot_true</span>, <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">2</span>)</pre></body></html></div>
<p><img src="twoDataInteraction_files/figure-html/unnamed-chunk-1-1.png" width="672"></p>
</div>
<div id="comparison-of-available-options-for-variable-selection" class="section level2">
<h2 class="hasAnchor">
<a href="#comparison-of-available-options-for-variable-selection" class="anchor"></a>Comparison of available options for variable selection</h2>
<p>SIComORe includes three alternative methods of selection for prediction of phenotype with grouped correlated structure, on top of the one defined in <span class="citation">Ambroise et al. (<a href="#ref-sicomore">2018</a>)</span>. MLGL (see <span class="citation">Grimonprez (<a href="#ref-grimonprez_PhD">2016</a>)</span>, <span class="citation">Grimonprez et al. (<a href="#ref-mlgl">2019</a>)</span>) and the variant <span class="math inline">\(\rho\)</span>-SIComORe in the spirit of <span class="citation">Park, Hastie, and Tibshirani (<a href="#ref-park">2007</a>)</span>. We try the 3 available options for simultaneously finding and selecting the groups of variables related to the phenotype.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">out_rho</span>      <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/sicomore.html">sicomore</a></span>(<span class="no">y</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">X1</span>, <span class="no">X2</span>), <span class="kw">selection</span><span class="kw">=</span><span class="st">"rho-sicomore"</span>)
<span class="no">out_mlgl</span>     <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/sicomore.html">sicomore</a></span>(<span class="no">y</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">X1</span>, <span class="no">X2</span>), <span class="kw">selection</span><span class="kw">=</span><span class="st">"mlgl"</span>        )
<span class="no">out_sicomore</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/sicomore.html">sicomore</a></span>(<span class="no">y</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">X1</span>, <span class="no">X2</span>), <span class="kw">selection</span><span class="kw">=</span><span class="st">"sicomore"</span>    )</pre></body></html></div>
<p>In this case, the three method tends to produce a similar estimate of the interaction coefficients:</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">plot_mlgl</span>     <span class="kw">&lt;-</span> <span class="no">out_mlgl</span>$<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">main</span> <span class="kw">=</span> <span class="st">"Estimated interactions (MLGL)"</span>, <span class="kw">threshold</span><span class="kw">=</span><span class="fl">.05</span>)
<span class="no">plot_sicomore</span> <span class="kw">&lt;-</span> <span class="no">out_sicomore</span>$<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">main</span> <span class="kw">=</span> <span class="st">"Estimated interactions (SICoMORE)"</span>, <span class="kw">threshold</span><span class="kw">=</span><span class="fl">.05</span>)
<span class="no">plot_rho</span>      <span class="kw">&lt;-</span> <span class="no">out_rho</span>$<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">main</span> <span class="kw">=</span> <span class="st">"Estimated interactions (rho-SICoMORE)"</span>, <span class="kw">threshold</span><span class="kw">=</span><span class="fl">.05</span>)
<span class="kw pkg">cowplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>( <span class="no">plot_mlgl</span>, <span class="no">plot_rho</span>,
                    <span class="no">plot_sicomore</span>, <span class="no">plot_true</span>, <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">2</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fl">2</span>)</pre></body></html></div>
<p><img src="twoDataInteraction_files/figure-html/plot%20comparison-1.png" width="672"></p>
<p>The method <code>getSignificance(effect = "beta")</code> sends back the results of a statistical test for each set of coefficients (main effect or interaction) in the form of <span class="math inline">\(1-p\)</span>-value. This can be used to determine at which extend a coefficient is selected (the closer from 1, the higher the probability for being non zero):</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">pval_mlgl</span>     <span class="kw">&lt;-</span> <span class="fl">1</span> - <span class="no">out_mlgl</span>$<span class="fu">getSignificance</span>(<span class="kw">effect</span> <span class="kw">=</span> <span class="st">"theta"</span>)
<span class="no">pval_rho</span>      <span class="kw">&lt;-</span> <span class="fl">1</span> - <span class="no">out_rho</span>$<span class="fu">getSignificance</span>(<span class="kw">effect</span> <span class="kw">=</span> <span class="st">"theta"</span>)
<span class="no">pval_sicomore</span> <span class="kw">&lt;-</span> <span class="fl">1</span> - <span class="no">out_sicomore</span>$<span class="fu">getSignificance</span>(<span class="kw">effect</span> <span class="kw">=</span> <span class="st">"theta"</span>)

<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">3</span>))
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="no">pval_mlgl</span>    , <span class="kw">main</span> <span class="kw">=</span> <span class="st">"p-values MLGL"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="no">pval_sicomore</span>, <span class="kw">main</span> <span class="kw">=</span> <span class="st">"p-values SIComORe"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="no">pval_rho</span>     , <span class="kw">main</span> <span class="kw">=</span> <span class="st">"p-values rho-SIComORe"</span>)</pre></body></html></div>
<p><img src="twoDataInteraction_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div id="comparison-of-the-computation-times-between-the-3-options" class="section level2">
<h2 class="hasAnchor">
<a href="#comparison-of-the-computation-times-between-the-3-options" class="anchor"></a>Comparison of the computation times between the 3 options</h2>
<p>As espected, the SICoMORe variants are much faster than the MLGL one.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="no">RHO</span>      <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span>(<span class="fu"><a href="../reference/sicomore.html">sicomore</a></span>(<span class="no">y</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">X1</span>, <span class="no">X2</span>), <span class="kw">selection</span><span class="kw">=</span><span class="st">"rho-sicomore"</span>, <span class="kw">verbose</span><span class="kw">=</span><span class="fl">FALSE</span>))
<span class="no">SICOMORE</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span>(<span class="fu"><a href="../reference/sicomore.html">sicomore</a></span>(<span class="no">y</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">X1</span>, <span class="no">X2</span>), <span class="kw">selection</span><span class="kw">=</span><span class="st">"sicomore"</span>    , <span class="kw">verbose</span><span class="kw">=</span><span class="fl">FALSE</span>))
<span class="no">MLGL</span>     <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span>(<span class="fu"><a href="../reference/sicomore.html">sicomore</a></span>(<span class="no">y</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">X1</span>, <span class="no">X2</span>), <span class="kw">selection</span><span class="kw">=</span><span class="st">"mlgl"</span>        , <span class="kw">verbose</span><span class="kw">=</span><span class="fl">FALSE</span>))
<span class="no">bench</span>    <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/microbenchmark/man/microbenchmark.html">microbenchmark</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/eval.html">eval</a></span>(<span class="no">RHO</span>), <span class="fu"><a href="https://rdrr.io/r/base/eval.html">eval</a></span>(<span class="no">SICOMORE</span>) , <span class="fu"><a href="https://rdrr.io/r/base/eval.html">eval</a></span>(<span class="no">MLGL</span>), <span class="kw">times</span> <span class="kw">=</span> <span class="fl">4</span>)</pre></body></html></div>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu">autoplot</span>(<span class="no">bench</span>)</pre></body></html></div>
<p><img src="twoDataInteraction_files/figure-html/timings%20plot-1.png" width="672"></p>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-sicomore">
<p>Ambroise, Christophe, Julien Chiquet, Florent Guinot, and Marie Szafranski. 2018. “Fast Computation of Genome-Metagenome Interaction Effects.” <em>arXiv Preprint arXiv:1810.12169</em>.</p>
</div>
<div id="ref-mlgl">
<p>Grimonprez, Q., S. Blanck, A Celisse, G. Marot, Y. Yang, and H. Zou. 2019. “MLGL: An R Package Implementing Correlated Variable Selection by Hierarchical Clustering and Group-Lasso.” <a href="https://cran.r-project.org/package=MLGL" class="uri">https://cran.r-project.org/package=MLGL</a>.</p>
</div>
<div id="ref-grimonprez_PhD">
<p>Grimonprez, Quentin. 2016. “Selection de Groupes de Variables Corrélées En Grande Dimension.” PhD thesis, Université de Lille; Lille 1.</p>
</div>
<div id="ref-park">
<p>Park, Mee Young, Trevor Hastie, and Robert Tibshirani. 2007. “Averaged Gene Expressions for Regression.” <em>Biostatistics</em> 8 (2): 212–27.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Christophe Ambroise, Julien Chiquet, Florent Guinot, Marie Szafranski.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
